Feat: Add fortnight support

linull24 opened the pull request:
Issue: #17 

 Added
- Support for fortnight (alternating week) courses
- Bottom-right components for class cards to display additional information
- Separate utility modules: `colorUtils.js` and `courseUtils.js`

 Changed
- Renamed mobile path from `/m` to `/mobile` for better code organization
- Refactored course data processing logic
- Optimized mobile interface display for better cross-device compatibility
- Updated documentation and API references

Imprortant
- Restructured `ScheduleTable.js` and `ReservedClassesList.js` mixins
- Updated state management in `store.js`
- Improved responsive layout for mobile views
![image](https://github.com/user-attachments/assets/8da84fd2-638c-46f1-8bb9-11cfc2902c13)
![image](https://github.com/user-attachments/assets/8db41b30-cb5f-4632-a2d9-e3ec7cf605d1)
![image](https://github.com/user-attachments/assets/688e4d2b-a60f-4d3b-98d3-64f80106a3ad)
![image](https://github.com/user-attachments/assets/3e857851-0173-4bff-bdff-24ec5a18580f)

linull24 committed:
Feat: Adding support for fortnight courses

linull24 committed:
Fix: Display fully on mobile

BH4HPA commented:
感谢您的贡献！从图片上看感觉重课课程的 UI 设计可能还需要商榷，现在这个样式仅仅可以体现这里有重课，但并不适合用作截图课表之类的场景（课程名称展示不全面），是否可以再调研下市面上其他课表软件如何处理重课的展示逻辑呢？比如对 PC 端来说，可以右下角放一个角标，hover 后展示双周课程之类的。Code Review 将在我有空的时候进行。

BH4HPA commented on src/utils/colorUtils.js:
文件名建议改一下，`colorUtils.js`已经放在`utils`文件夹下面了，不需要再带`Utils`字样。

```import * as ColorUtils from 'src/utils/color'```
is better than
```import * as ColorUtils from 'src/utils/colorUtils'```

BH4HPA commented on src/utils/courseUtils.js:
同 `colorUtils` 的问题，建议改下文件名

BH4HPA commented on src/utils/courseUtils.js:
```diff
@@ -0,0 +1,155 @@
+
+              const currentWeekType = period[4]; 
+              
+              // 检查是否有真正的冲突（考虑单双周）
```
「真正的冲突」不宜在代码中体现，建议`realConflict`变量改名为`confilct`或者`conflicted`

BH4HPA commented on src/utils/courseUtils.js:
```diff
@@ -0,0 +1,155 @@
+                      (currentWeekType === '双' && conflictWeekType === '单')) {
+                    continue;
+                  } else {
+
```
空行

BH4HPA commented on src/utils/courseUtils.js:
```diff
@@ -0,0 +1,155 @@
+      if (reserved.hasOwnProperty(currentRowsMap[key]['course_id']) && reserved[currentRowsMap[key]['course_id']].classes.hasOwnProperty(currentRowsMap[key]['teacher_id'])) {
+        if (rowsMap.hasOwnProperty(key)) {
+          if (selected.hasOwnProperty(currentRowsMap[key]['course_id']) && selected[currentRowsMap[key]['course_id']].teacherId === currentRowsMap[key]['teacher_id'] && currentRowsMap[key]['class_time'] !== rowsMap[key]['class_time']) {
+            getPeriods(rowsMap[key]['class_time']).forEach((period) => {
```
这里的 `period` 考虑使用数组析构作入参，因为下面的 `period[0]` `period[1]` 都语义不明。

```
第(0-12)节课, 周(0-4), 是否为时段第一节, periods的长度, 单双周信息, 是否为实验课
```

参考

```
getPeriods(rowsMap[key]['class_time']).forEach(([startPeriod, day, _, periodLength, weekType, isExperiment]) => {
```

BH4HPA commented on src/utils/courseUtils.js:
```diff
@@ -0,0 +1,155 @@
+                const conflictClassTime = reserved[conflictCourseId].classes[conflictTeacherId].classTime;
+                const conflictPeriods = getPeriods(conflictClassTime);
+                const conflictPeriod = conflictPeriods.find(p => 
+                  p[0] === period[0] && p[1] === period[1]
```
是否考虑支持了单周三节课双周两节课这种重合的情况？

BH4HPA commented on src/utils/courseUtils.js:
```diff
@@ -0,0 +1,155 @@
+  });
+}
+
+export const isMacLike = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
```
这个不适合放在 `CourseUtils` 里

BH4HPA commented on README-zh_CN.md:
```diff
@@ -82,8 +87,10 @@ $ yarn build
 │   ├── storage.js   # 处理数据压缩与持久化
-│   ├── store.js   # 数据处理的核心部分
-│   └── utils.js   # 一些可复用的逻辑
+│   ├── store.js   # Vuex状态管理
```
和英文版不对应

BH4HPA commented on README-zh_CN.md:
```diff
@@ -67,7 +72,7 @@ $ yarn build
 │   │   │   │   └── ......
 │   │   │   └── main.js
-│   │   ├── m   # 移动版页面及组件
+│   │   ├── mobile   # 移动版
```
这里的注释修改有歧义，移动版和桌面版还是有部分组件是复用的

BH4HPA commented on docs/api-zh_CN.md:
```diff
@@ -309,19 +309,40 @@ ec17f6e2:00000000+1002,00000000-1001
+    []
+  ],
+  [[], [], [], [], []],
+  [[], [], [], [], []]
```
这里例子可以补一个地方放两节课

BH4HPA commented on src/mixins/ReservedClassesList.js:
```diff
@@ -51,28 +52,95 @@ export const ReservedClassesListMixin = {
-                  let cell = this.$store.getters.scheduleTableRows[period[0]][period[1]];
-                  if (cell !== null && cell.courseId !== key) {
-                    isConflicted = true;
+                  let cells = this.$store.getters.scheduleTableRows[period[0]][period[1]];
```
这里没有复用 utils 里写好的逻辑，是否可以考虑改一下 CourseUtils 里的函数让他在这里也可以被复用？

BH4HPA commented on src/mixins/ReservedClassesList.js:
```diff
@@ -51,28 +52,95 @@ export const ReservedClassesListMixin = {
-                      if (cellBefore != null && cellBefore.campus !== this.$store.state.reservedClasses[key].classes[teacherId].campus
-                        && cellBefore.courseId !== this.id) {
-                        isConflicted = true;
+
```
重复的代码片段很多

BH4HPA commented on src/mixins/ReservedClassesList.js:
```diff
@@ -148,23 +216,96 @@ export const CourseClassesListMixin = {
-              if (cellAfter != null && cellAfter.campus !== this.$store.state.reservedClasses[this.id].classes[key].campus
-                && cellAfter.courseId !== this.id && courseConflicts[cellAfter.courseId] == null) {
-                courseConflicts[cellAfter.courseId] = 2;
+        const currentClassPeriods = getPeriods(this.$store.state.reservedClasses[this.id].classes[key].classTime);
```
重复的代码片段很多

BH4HPA commented on src/mixins/ReservedClassesList.js:
```diff
@@ -148,23 +216,96 @@ export const CourseClassesListMixin = {
+                  courseConflicts[cell.courseId] = 1;
+                }
               }
+            });
```
重复的代码片段很多

BH4HPA commented on src/mixins/ReservedClassesList.js:
```diff
@@ -51,28 +52,95 @@ export const ReservedClassesListMixin = {
-                      if (cellAfter != null && cellAfter.campus !== this.$store.state.reservedClasses[key].classes[teacherId].campus
-                        && cellAfter.courseId !== this.id) {
-                        isConflicted = true;
+
```
重复的代码片段很多

BH4HPA commented on src/mixins/ScheduleTable.js:
```diff
@@ -222,7 +266,195 @@ export const ClassCardMixin = {
+      await this.adjustAllTextElements();
     },
+
+    async adjustAllTextElements() {
```
这玩意没有 css 实现的方案吗，不建议拿 js 高频操作样式

BH4HPA commented on src/pages/index/components/ScheduleTable/NoPeriodClassCard.bottom-right.vue:
```diff
@@ -0,0 +1,69 @@
+</template>
+
+<script>
+  import {ClassCardMixin} from '../../../../mixins/ScheduleTable';
```
删前导空格，{}内补空格，`../../../..`推导链太长是否可以`@`或者`/src`起始搜索

BH4HPA commented on src/pages/index/components/ScheduleTable/index.vue:
```diff
@@ -45,11 +45,29 @@
+              <div class="course-stack" v-if="courses.length > 0">
+                <template v-for="(course, courseIndex) in courses">
+                  <ClassCard.bottom-right 
+                    v-if="course != null && !course.qr && course.clipPathMode === 'bottom-right'"
```
组件 `ClassCard.bottom-right` 和 `ClassCard` 传参基本上是一样的，可以考虑`ClassCard`改名`ClassCard.Single`，`ClassCard`改名`ClassCard.Double`然后再开一个组件`ClassCard`，把使用 `Single` 还是 `Double` 的逻辑放在新开的组件里，这样外部就不需要做判断了。

BH4HPA commented on src/pages/mobile/components/ScheduleTable/index.vue:
```diff
@@ -10,10 +10,27 @@
+              v-if="courses.length === 0 || (courses.length > 0 && courses[0] != null && courses[0].first)">
+            <div class="course-stack" v-if="courses.length > 0">
+              <template v-for="(course, courseIndex) in courses">
+                <ClassCard.bottom-right 
```
同桌面端

BH4HPA commented on src/workers/filter.js:
```diff
@@ -81,22 +81,117 @@ registerPromiseWorker(function (message) {
+              continue;
+            }
+            
+            const conflictTeacherId = message.selectedClasses[conflictCourseId].teacherId;
```
这个重复的代码段又出现了

BH4HPA commented on src/workers/filter.js:
```diff
@@ -81,22 +81,117 @@ registerPromiseWorker(function (message) {
+                  continue;
+                }
+                
+                const conflictTeacherId = message.selectedClasses[conflictCourseId].teacherId;
```
重复的代码段

BH4HPA commented on src/workers/filter.js:
```diff
@@ -81,22 +81,117 @@ registerPromiseWorker(function (message) {
+                  continue;
+                }
+                
+                const conflictTeacherId = message.selectedClasses[conflictCourseId].teacherId;
```
重复的代码段

linull24 replied:
我打算这里新开扔到utils里面去

BH4HPA commented on yarn.lock:
`yarn.lock` 怎么改了这么多东西，试一下恢复这个文件的更改

linull24 replied:
一开始我直接裸机npm23yarn install过 后来根据提示，后面都是nvm用npm 17。可能是这里导致了问题

linull24 commented:
UI部分的话我再改改字体处理和多边形显示，可能改为直角梯形...因为目前市面上的要么干脆根据当前周显示，要么直接挤在一起。我一开始选择三角形就是能够直观的显示，左上为单周 右下为双周
btw，我以前对于前端的理解还在jquery和ajax php（捂脸，这几天嗯写了不少vue2，的确体会到现代web，虽然vue2也过时了...
这个项目，我估计迟早得转到vue3的轮子上去，但是目前肯定是得先保证能跑

linull24 committed:
Rename : Remove "XXXUtils"

linull24 committed:
Fixed to meet maintainer's requirement:
1.  改文档
2.  分离矛盾判断处理逻辑
3.  重命名utils下文件
4.  分离字体大小调整（css据我所知不能严格实现）
5.  增添对于字体字体大小调整docs/AdjustTextSize.md的说明
6.  回滚yarn.lock
7.  合并classcard

linull24 commented:
![image](https://github.com/user-attachments/assets/9a0891ff-6d64-4dbc-993c-1b41698d5386)
我只能说在这么大点地方，直观 放下两节课的信息，除此以外的非交互方案，只有横着切，但是也不直观，当然你也可以自己调参数。毕竟为了截图这是最后的妥协方案。

BH4HPA reviewed (approved):
个人感觉这个 AdjustTextSize 感觉还是不太好

linull24 commented:
> 个人感觉这个 AdjustTextSize 感觉还是不太好
1. CSS 无法测量渲染后的实际尺寸
2. CSS对于多边形的ClipPath支持不足
3. 分离了智能缩写这部分原来在coursetable的部分，也应该分离。
4. CSS不能保证课程名的优先级。

ZKLlab commented on src/mixins/ReservedClassesList.js:
```diff
@@ -254,5 +193,17 @@ export const CourseMetaMixin = {
         return null;
       }
     },
+    getCourseInfo() {
```
这个计算属性和 GetCoursePropertyMixin 里的 method 重名了，在 CourseMeta 组件里两者会冲突

ZKLlab commented on src/pages/index/components/ScheduleTable/ClassCard.Double.vue:
```diff
@@ -0,0 +1,103 @@
+    <div class="extra" ref="extraRef" :style="style[theme][1]" v-if="course.lab">
+      {{ course.lab }}
+    </div>
+    <div class="venue" ref="venueRef" :style="style[theme][1]" v-else>
```
上课地点应该在 props 的 venue 传入 true 时才展示，和教师互斥

ZKLlab commented on src/utils/course.js:
```diff
@@ -0,0 +1,129 @@
+  });
+}
+
+export function isMacLike() {
```
isMacLike 原来是个常量，如果改成函数，那使用的地方都要改成函数调用

ZKLlab commented on src/workers/filter.js:
```diff
@@ -120,10 +173,10 @@ registerPromiseWorker(function (message) {
           conditionsRegExp[condition] = new RegExp(message.conditions.search[condition], 'i');
         } catch (e) {
-          conditionsRegExp[condition] = concatRegExp(message.conditions.search[condition].split(/\s+/));
+          conditionsRegExp[condition] = concatRegExp(message.conditions.search[condition].split(/\\s+/));
```
这里原本的作用是将搜索词里的空白字符转换为匹配任意字符，例如搜索“概率论 数理统计”可以匹配“概率论与数理统计”，为什么要将 `\s+` 改成 `\\s+` 呢？下同

ZKLlab commented on src/workers/filter.js:
```diff
@@ -1,10 +1,10 @@
 function concatRegExp (parts) {
   parts.forEach((part, index) => {
-    parts[index] = part.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+    parts[index] = part.replace(/[.*+?^${}()|[\\\\]\\\\\\\\]/g, '\\\\\\\\$&');
```
这么多反斜杠是？

ZKLlab reviewed (commented):
用 `clamp(12px, 1.2vw, 14px)` 的写法来写课表内文字尺寸有几个问题，建议再考虑一下实现方式：

1. 课表宽度不一定和视口宽度成比例，当桌面端视口宽度超过一个值时，课表宽度会变成固定值，此时课表内的文字仍然在跟随视口宽度变化大小，会显得比较奇怪并且像 bug
2. 现在有用 JS 计算文字大小的逻辑，同时有 CSS 控制字体大小会使逻辑变得耦合和复杂，建议只保留一种字体大小控制方式
3. 在一般解析度的屏幕使用非整数字体大小，会让文字模糊或偏移行中心，渲染效果不佳

ZKLlab commented on src/utils/AdjustTextSize.js:
```diff
@@ -0,0 +1,260 @@
+      whiteSpace: '', overflow: '', textOverflow: '', 
+      zIndex: '', outline: ''
+    });
+    await vue.$nextTick();
```
应该是不用 $nextTick 的，下面 window.getComputedStyle 的时候会同步回流重绘，下同

ZKLlab commented on src/mixins/ScheduleTable.js:
```diff
@@ -206,6 +256,7 @@ export const ClassCardMixin = {
         clearTimeout(this.timer);
         this.timer = setTimeout(async () => {
           await this.doShortenCourseName();
+          await this.adjustAllTextElements();
```
doShortenCourseName 里不是调了一次 adjustAllTextElements 吗？

ZKLlab reviewed (commented):
异形容器内可以用 [shape-outside](https://developer.mozilla.org/zh-CN/docs/Web/CSS/shape-outside) + float 实现文本排版适应形状，避免文本被切到

![image](https://github.com/user-attachments/assets/325b3643-cc48-4ffc-a0ae-e74b699b9619)

ZKLlab commented on src/utils/AdjustTextSize.js:
```diff
@@ -0,0 +1,260 @@
+        
+        if (hasShortLine || text.length <= minChars) {
+          Object.assign(el.style, {
+            whiteSpace: 'nowrap',
```
感觉这里的 nowrap 影响了缩短课程名的逻辑，考虑执行缩短课程名逻辑前是否应该重置样式

表现是调整视口宽度过程中，能看到部分课程名在“有缩短-无缩短”之间来回变化

https://github.com/user-attachments/assets/cfd3d24d-2ca6-4f87-baa6-54de7673dd06

BH4HPA commented:
转二学期之后，可能一个时段会选上两节课(1-8周)、(9-16周），是否考虑本次也支持这个功能？考虑把上课周解析成 weekPattern 的形式去做冲突检测？

```ts
Array.from({ length: 20 }, (_, idx) => idx + 1).filter(week => weekPattern & (1 << (week - 1))),
```

https://github.com/BH4HPA/shu-charging-status/blob/f916643475a695b9a45ab2dcdd686c470566b87b/src/routes/shu-calendar.tsx#L789-L844

```tsx
          <Box sx={{ mt: 2, mb: 1, display: 'flex', justifyContent: 'center', alignItems: 'center' }}>
            <Typography variant='body1' sx={{ flexGrow: 1 }}>
              生效周
            </Typography>
            <Button size="small"
              onClick={() => setWeekPattern(prev => prev === 0b1111111111 ? 0 : 0b1111111111)}
            >
              {weekPattern !== 0b1111111111 ? '全选' : '全不选'}
            </Button>
            <Button size="small"
              onClick={() => setWeekPattern(0b0101010101)}
            >
              单周
            </Button>
            <Button size="small"
              onClick={() => setWeekPattern(0b1010101010)}
            >
              双周
            </Button>
          </Box>
          <Box
            sx={{
              display: 'flex',
              justifyContent: 'space-between',
              alignItems: 'center',
              mb: 2
            }}
          >
            {
              weekPattern.toString(2).padStart(10, '0').split('').reverse().map((week, idx) => (
                <Box
                  key={idx}
                  sx={{
                    display: 'flex',
                    flexDirection: 'column',
                    alignItems: 'center',
                    justifyContent: 'center',
                  }}
                >
                  <Checkbox
                    size='small'
                    sx={{ p: 0 }}
                    checked={week === '1'}
                    onChange={(e) => {
                      if (e.target.checked) {
                        setWeekPattern(weekPattern | (1 << (idx)));
                      } else {
                        setWeekPattern(weekPattern & ~(1 << (idx)));
                      }
                    }}
                  />
                  <Typography variant='body2'>{idx + 1}</Typography>
                </Box>
              ))
            }
          </Box>
```

---

https://github.com/BH4HPA/shu-course-calendar/blob/0a6972b4e9d3fb56cb5bd3da957d99aae548ddde/src/parser.ts#L163-L243

https://github.com/BH4HPA/shu-course-calendar/blob/0a6972b4e9d3fb56cb5bd3da957d99aae548ddde/src/parser.ts#L272-L281

linull24 commented:
> 转二学期之后，可能一个时段会选上两节课(1-8周)、(9-16周），是否考虑本次也支持这个功能？考虑把上课周解析成 weekPattern 的形式去做冲突检测？
> 
> ```ts
> Array.from({ length: 20 }, (_, idx) => idx + 1).filter(week => weekPattern & (1 << (week - 1))),
> ```
> 
> https://github.com/BH4HPA/shu-charging-status/blob/f916643475a695b9a45ab2dcdd686c470566b87b/src/routes/shu-calendar.tsx#L789-L844
> 
> ```tsx
>           <Box sx={{ mt: 2, mb: 1, display: 'flex', justifyContent: 'center', alignItems: 'center' }}>
>             <Typography variant='body1' sx={{ flexGrow: 1 }}>
>               生效周
>             </Typography>
>             <Button size="small"
>               onClick={() => setWeekPattern(prev => prev === 0b1111111111 ? 0 : 0b1111111111)}
>             >
>               {weekPattern !== 0b1111111111 ? '全选' : '全不选'}
>             </Button>
>             <Button size="small"
>               onClick={() => setWeekPattern(0b0101010101)}
>             >
>               单周
>             </Button>
>             <Button size="small"
>               onClick={() => setWeekPattern(0b1010101010)}
>             >
>               双周
>             </Button>
>           </Box>
>           <Box
>             sx={{
>               display: 'flex',
>               justifyContent: 'space-between',
>               alignItems: 'center',
>               mb: 2
>             }}
>           >
>             {
>               weekPattern.toString(2).padStart(10, '0').split('').reverse().map((week, idx) => (
>                 <Box
>                   key={idx}
>                   sx={{
>                     display: 'flex',
>                     flexDirection: 'column',
>                     alignItems: 'center',
>                     justifyContent: 'center',
>                   }}
>                 >
>                   <Checkbox
>                     size='small'
>                     sx={{ p: 0 }}
>                     checked={week === '1'}
>                     onChange={(e) => {
>                       if (e.target.checked) {
>                         setWeekPattern(weekPattern | (1 << (idx)));
>                       } else {
>                         setWeekPattern(weekPattern & ~(1 << (idx)));
>                       }
>                     }}
>                   />
>                   <Typography variant='body2'>{idx + 1}</Typography>
>                 </Box>
>               ))
>             }
>           </Box>
> ```
> 
> https://github.com/BH4HPA/shu-course-calendar/blob/0a6972b4e9d3fb56cb5bd3da957d99aae548ddde/src/parser.ts#L163-L243
> 
> https://github.com/BH4HPA/shu-course-calendar/blob/0a6972b4e9d3fb56cb5bd3da957d99aae548ddde/src/parser.ts#L272-L281

最近忙着期末，看情况，先根据zkl学长的修改来；
至于二学期，我觉得可能需要重新fork开发或者说是等我的pr足够merge以后再来。当初我开发的时候没有考虑到学期改制。
以我的理解来说，如果考虑单双+小学期课，那么一个课表显示是不可能的。那么只能虚拟——拆分为两个表显示上半学期和下半学期。
另外，新选课系统会上线，目前新系统在测试阶段。以我个人精力，很难说是在新夏季上线前完成了。
